### PR TITLE
Implement Fish Mocks

### PR DESCRIPTION
**Pull request by thelooter**
# Description
Implement Fish Mocks

# Checklist
The following items should be checked before the pull request can be merged.
- [x] Unit tests added.
- [x] Code follows existing code format.

# Info on creating a pull request
- Make sure that unit tests are added which test the relevant changes.
- Make sure that the changes follow the existing code format.

# For maintainer
When a PR is approved, the maintainer should label this PR with `release/*`.

- If the PR fixes a bug in MockBukkit, update the patch version. (if the current version is `0.4.1`, the new version should be `0.4.2`)
- If the PR adds a new feature to MockBukkit, update minor version. (if the current version is `0.4.1`, the new version should be `0.5.0`)

Note that a PR that fixes an `UnimplementedOperationException` should be considered a new version and not a bugfix.

### CODE REVIEW

File: src/main/java/be/seeseemelk/mockbukkit/entity/FishMock.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+	public void setFromBucket(boolean fromBucket)
+	{
+		this.isFromBucket = fromBucket;
+	}
+
+	@Override
+	public abstract @NotNull ItemStack getBaseBucketItem();
+
+	@Override
+	public @NotNull Sound getPickupSound(){
```
Comment (Insprill) [inline src/main/java/be/seeseemelk/mockbukkit/entity/FishMock.java]:
```suggestion
	public @NotNull Sound getPickupSound() {

File: src/main/java/be/seeseemelk/mockbukkit/entity/PufferFishMock.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+	@Override
+	public int getPuffState()
+	{
+		return this.puffState;
+	}
+
+	@Override
+	public void setPuffState(int state)
+	{
+		Preconditions.checkArgument((0 <= state) && (state <= 2), "State has to be between 0 and 2");
```
Comment (Insprill) [inline src/main/java/be/seeseemelk/mockbukkit/entity/PufferFishMock.java]:
```suggestion
		Preconditions.checkArgument((state >= 0) && (state <= 2), "State has to be between 0 and 2");

File: src/main/java/be/seeseemelk/mockbukkit/entity/TadpoleMock.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+	@Override
+	public int getAge()
+	{
+		return this.age;
+	}
+
+	@Override
+	public void setAge(int age)
+	{
+		Preconditions.checkArgument(age < 24000 || (this.age + age) < 24000, "Tadpole age can't be greater than 24000");
```
Comment (Insprill) [inline src/main/java/be/seeseemelk/mockbukkit/entity/TadpoleMock.java]:
```suggestion
		Preconditions.checkArgument(this.age + age < 24000, "Tadpole age can't be greater than 24000");

File: src/main/java/be/seeseemelk/mockbukkit/entity/TropicalFishMock.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+import java.util.UUID;
+import java.util.concurrent.ThreadLocalRandom;
+
+public class TropicalFishMock extends FishMock implements TropicalFish
+{
+
+	private @NotNull DyeColor patternColor;
+	private @NotNull DyeColor bodyColor;
+	private @NotNull Pattern pattern;
+
```
Comment (Insprill) [inline src/main/java/be/seeseemelk/mockbukkit/entity/TropicalFishMock.java:21]:
```suggestion

File: src/main/java/be/seeseemelk/mockbukkit/entity/TropicalFishMock.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+	public TropicalFishMock(@NotNull ServerMock server, @NotNull UUID uuid)
+	{
+
```
Comment (Insprill) [inline src/main/java/be/seeseemelk/mockbukkit/entity/TropicalFishMock.java]:
```suggestion

File: src/test/java/be/seeseemelk/mockbukkit/entity/PufferFishMockTest.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+	void testGetPuffStateDefault()
+	{
+		assertEquals(0, pufferFish.getPuffState());
+	}
+
+	@Test
+	void testSetPuffState()
+	{
+		pufferFish.setPuffState(1);
+		assertEquals(1,pufferFish.getPuffState());
```
Comment (Insprill) [inline src/test/java/be/seeseemelk/mockbukkit/entity/PufferFishMockTest.java]:
```suggestion
		assertEquals(1, pufferFish.getPuffState());

File: src/test/java/be/seeseemelk/mockbukkit/entity/SalmonMockTest.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+	{
+		assertEquals(Material.SALMON_BUCKET, salmon.getBaseBucketItem().getType());
+	}
+
+	@Test
+	void testGetType()
+	{
+		assertEquals(EntityType.SALMON, salmon.getType());
+	}
+}
```
Comment (Insprill) [inline src/test/java/be/seeseemelk/mockbukkit/entity/SalmonMockTest.java]:
```suggestion
	}

}

File: src/test/java/be/seeseemelk/mockbukkit/entity/TadpoleMockTest.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+	void testGetAge()
+	{
+		assertEquals(0,tadpole.getAge());
+	}
+
+	@Test
+	void testSetAge()
+	{
+		tadpole.setAge(15000);
+		assertEquals(15000,tadpole.getAge());
```
Comment (Insprill) [inline src/test/java/be/seeseemelk/mockbukkit/entity/TadpoleMockTest.java]:
```suggestion
		assertEquals(15000, tadpole.getAge());

File: src/test/java/be/seeseemelk/mockbukkit/entity/TadpoleMockTest.java [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+		assertThrows(IllegalArgumentException.class, () -> tadpole.setAge(426900));
+	}
+
+	@Test
+	void testSetAgeToBigWithPreviousValue()
+	{
+		tadpole.setAge(15000);
+		assertThrows(IllegalArgumentException.class, () -> tadpole.setAge(426900));
+	}
+}
```
Comment (Insprill) [inline src/test/java/be/seeseemelk/mockbukkit/entity/TadpoleMockTest.java]:
```suggestion
	}

}

File: src/test/java/be/seeseemelk/mockbukkit/entity/TadpoleMockTest.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+	void testSetAgeToBig()
+	{
+		assertThrows(IllegalArgumentException.class, () -> tadpole.setAge(426900));
+	}
+
+	@Test
+	void testSetAgeToBigWithPreviousValue()
+	{
+		tadpole.setAge(15000);
+		assertThrows(IllegalArgumentException.class, () -> tadpole.setAge(426900));
```
Comment (Insprill) [inline src/test/java/be/seeseemelk/mockbukkit/entity/TadpoleMockTest.java:69]:
```suggestion
		tadpole.setAge(23000);
		assertThrows(IllegalArgumentException.class, () -> tadpole.setAge(2000));

File: src/test/java/be/seeseemelk/mockbukkit/entity/TropicalFishMockTest.java [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+	void testGetPatternColor()
+	{
+		assertTrue(Arrays.stream(DyeColor.values()).anyMatch(dyeColor -> tropicalFish.getPatternColor().equals(dyeColor)));
+	}
+
+	@Test
+	void testSetPatternColor()
+	{
+		tropicalFish.setPatternColor(DyeColor.RED);
+		assertEquals(DyeColor.RED,tropicalFish.getPatternColor());
```
Comment (Insprill) [inline src/test/java/be/seeseemelk/mockbukkit/entity/TropicalFishMockTest.java]:
```suggestion
		assertEquals(DyeColor.RED, tropicalFish.getPatternColor());

File: src/test/java/be/seeseemelk/mockbukkit/entity/TropicalFishMockTest.java [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+	}
+
+	@Test
+	void testGetBodyColor()
+	{
+		assertTrue(Arrays.stream(DyeColor.values()).anyMatch(dyeColor -> tropicalFish.getBodyColor().equals(dyeColor)));
```
Comment (Insprill) [inline src/test/java/be/seeseemelk/mockbukkit/entity/TropicalFishMockTest.java]: I don't think this is possible to fail. Perhaps a simple notNull check would be better?

File: src/test/java/be/seeseemelk/mockbukkit/entity/TropicalFishMockTest.java [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+	void testSetBodyColor()
+	{
+		tropicalFish.setBodyColor(DyeColor.RED);
+		assertEquals(DyeColor.RED,tropicalFish.getBodyColor());
+	}
+
+	@Test
+	void testGetPattern()
+	{
+		assertTrue(Arrays.stream(Pattern.values()).anyMatch(pattern -> tropicalFish.getPattern().equals(pattern)));
```
Comment (Insprill) [inline src/test/java/be/seeseemelk/mockbukkit/entity/TropicalFishMockTest.java]: I don't think this is possible to fail. Perhaps a simple notNull check would be better?

File: src/test/java/be/seeseemelk/mockbukkit/entity/TropicalFishMockTest.java [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+	}
+
+	@Test
+	void testSetPattern()
+	{
+		tropicalFish.setPattern(Pattern.BETTY);
+		assertEquals(Pattern.BETTY, tropicalFish.getPattern());
+	}
+}
```
Comment (Insprill) [inline src/test/java/be/seeseemelk/mockbukkit/entity/TropicalFishMockTest.java]:
```suggestion
	}

}

File: src/main/java/be/seeseemelk/mockbukkit/entity/TropicalFishMock.java [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+		super(server, uuid);
+
+		System.out.println(DyeColor.values().length);
+		System.out.println(Pattern.values().length);
```
Comment (Insprill) [inline src/main/java/be/seeseemelk/mockbukkit/entity/TropicalFishMock.java]:
```suggestion

**Review by Insprill — CHANGES_REQUESTED** [review]

**Review by Insprill — CHANGES_REQUESTED** [review]
Just missing the logic to actually spawn them :P

**Comment by thelooter** [comment]
Yeah working on it

File: src/test/java/be/seeseemelk/mockbukkit/WorldMockTest.java [hunk 16]
```diff
@@ -1155,4 +1160,48 @@ void testSpawnFox()
...
+
+	@Test
+	void testSpawnPufferfish()
+	{
+		WorldMock world = new WorldMock(Material.DIRT, 3);
+		Entity entity = world.spawnEntity(new Location(world, 0, 0, 0), EntityType.PUFFERFISH);
+		assertInstanceOf(PufferFishMock.class, entity);
+		assertTrue(entity.isValid());
+	}
 }
```
Comment (Insprill) [inline src/test/java/be/seeseemelk/mockbukkit/WorldMockTest.java]:
```suggestion
	}

}

**Review by Insprill — CHANGES_REQUESTED** [review]

**Review by Insprill — APPROVED** [review]
LGTM

**Pull request merged**